src/dos/dos_keyboard_layout.cpp

```diff
@@ -958,306 +1021,269 @@ KeyboardErrorCode KeyboardLayout::ReadCodePageFile(const char *requested_cp_file
 				font_data_start+=font_height*256;
 			}
 
-			LOG(LOG_BIOS,LOG_NORMAL)("Codepage %i successfully loaded",codepage_id);
-
-			// set codepage entries
-			dos.loaded_codepage=(uint16_t)(codepage_id&0xffff);
+			// Set code page entries
+			dos.loaded_codepage   = code_page;
+			code_page_font_origin = cpi_file.empty()
+			                              ? CodePageFontOrigin::Bundled
+			                              : CodePageFontOrigin::Custom;
 
 			// update font if necessary
 			if (font_changed && (CurMode->type==M_TEXT) && (IS_EGAVGA_ARCH)) {
 				INT10_ReloadFont();
 			}
 			INT10_SetupRomMemoryChecksum();
 
-			notify_code_page_changed();
-			return KEYB_NOERROR;
+			return KeyboardLayoutResult::OK;
 		}
 
 		start_pos = host_readd(&cpi_buf.at(start_pos));
 		start_pos+=2;
 	}
 
-	LOG(LOG_BIOS,LOG_ERROR)("Codepage %i not found",codepage_id);
-
-	return KEYB_INVALIDCPFILE;
+	if (cpi_file.empty()) {
+		LOG_WARNING("DOS: Code page %d not found in a bundled CPI file",
+		            code_page);
+		return KeyboardLayoutResult::NoBundledCpiFileForCodePage;
+	}
+	return KeyboardLayoutResult::NoCodePageInCpiFile;
 }
 
-bool KeyboardLayout::HasLanguageCode(const char *requested_code)
+bool KeyboardLayout::IsLayoutAvailable(const std::string& keyboard_layout)
 {
-	for (const auto &language_code : language_codes)
-		if (iequals(language_code, requested_code))
+	for (const auto& availalbe_layout : available_layouts) {
+		if (iequals(availalbe_layout, keyboard_layout)) {
 			return true;
+		}
+	}
+
 	return false;
 }
 
-KeyboardErrorCode KeyboardLayout::SwitchKeyboardLayout(const char *new_layout,
-                                                       KeyboardLayout *&created_layout,
-                                                       int32_t &tried_cp)
+KeyboardLayoutResult KeyboardLayout::SwitchKeyboardLayout(
+        const std::string& keyboard_layout, KeyboardLayout*& created_layout)
 {
-	assert(new_layout);
-
-	if (!iequals(new_layout, "US")) {
+	if (!iequals(keyboard_layout, "US")) {
 		// switch to a foreign layout
 
-		if (HasLanguageCode(new_layout)) {
+		if (IsLayoutAvailable(keyboard_layout)) {
 			if (!use_foreign_layout) {
 				// switch to foreign layout
-				use_foreign_layout  = true;
-				diacritics_character=0;
-				LOG(LOG_BIOS, LOG_NORMAL)("Switched to layout %s", new_layout);
+				use_foreign_layout   = true;
+				diacritics_character = 0;
 			}
 		} else {
 			KeyboardLayout *temp_layout = new KeyboardLayout();
-			auto req_codepage           = temp_layout->ExtractCodePage(new_layout);
-			tried_cp                    = req_codepage;
-			auto rcode                  = temp_layout->ReadKeyboardFile(new_layout, req_codepage);
-			if (rcode) {
+
+			auto code_page = temp_layout->ExtractCodePage(keyboard_layout);
+			if (!IS_EGAVGA_ARCH && code_page != DefaultCodePage) {
```

Comment:
I'd prefer `(machine==MCH_EGA) || (machine==MCH_VGA)`.

Reply:
Changed in the whole file.